### DISCUSSION TITLE
Custom loader not called on node 20

### DISCUSSION THREAD

**Discussion by enricoschaaf** [avajs/ava#3213] [Q&A]
Please provide details about:

* What you're trying to do
Load a typescript test file
* What happened
* TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for example.test.ts
* What you expected to happen
the typescript loader to be called and handeling the file

Please share relevant sample code. Or better yet, provide a link to a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example).

We'll also need your AVA configuration (in `package.json` or `ava.config.*` configuration files) and how you're invoking AVA. Share the installed AVA version (get it by running `npx ava --version`).

  "ava": {
    "extensions": {
      "ts": "module"
    },
    "nodeArguments": [
      "--loader=tsx"
    ]
  }
  
  node version 20.3.0 ava verison 5.3.0

**Comment by novemberborn** [comment]
I reckon the problem is somewhere between Node.js, the loader and the configuration; not AVA itself.

One possibility is that with your configuration (`"ts": "module"`) the test file is loaded through `import()` not `require()`.  The loader may be CJS only. Try `"ts": "commonjs"` .

**Comment by transitive-bullshit** [comment]
Running into the same issue which only repros on Node.js v20.

Here's a [workaround](https://github.com/avajs/ava/issues/2593#issuecomment-1524846453), but something changed between Node v19 and v20 which is causing this.

**Reply by novemberborn** [reply]
Looks like the argument needs to be `--experimental-loader`: https://nodejs.org/api/esm.html#loaders

One thing to keep in mind, AVA's `nodeArguments` is used when starting a worker thread or launching a child process. It _could_ be the case that you (no longer?) cannot add new loaders in those circumstances, which is why that `NODE_OPTIONS` configuration is required.

But hey, these are _experimental_ features after all.

**Reply by tommy-mitchell** [reply]
Maybe a new `loaders` option could be added to AVA’s configuration? In Node 20, it could add the loaders to the `NODE_OPTIONS` environment variable, and in prior versions it could add them to the current `nodeArguments` option:

```json

{
	"ava": {
		"loaders": [
			"ts-node/esm",
			"esmock"
		],
		"nodeArguments": [
			"--no-warnings=ExperimentalWarnings"
		]
	}
}
```

**Reply by novemberborn** [reply]
Not until this is no longer experimental in Node.js itself.

I don't think we can even install loaders dynamically in Node.js 20 at this point, modifying `NODE_OPTIONS` would have no impact on worker threads.

**Reply by tommy-mitchell** [reply]
https://github.com/nodejs/node/pull/46826 (merged but unreleased) adds support for programmatically registering loaders.

**Reply by tommy-mitchell** [reply]
Looks like it’s still a while away, though: https://github.com/nodejs/loaders/issues/153#issuecomment-1656350600

**Reply by novemberborn** [reply]
Once that's in a Node.js version, we could make it an experimental feature in AVA. This means it needs to be explicitly opted in to (and we make no reliability guarantees). But then once it's stable in Node.js it's a straight-forward change to make it stable in AVA as well.

**Reply by tommy-mitchell** [reply]
Looks like it's coming in Node.js 20.6: https://github.com/nodejs/node/pull/49185

**Reply by tommy-mitchell** [reply]
So this has since released. How are experimental features in AVA developed?

**Reply by tommy-mitchell** [reply]
@novemberborn friendly bump :) how should we go about adding this feature? (Or is there a better way?) Seems like it's primarily [`register(loader)` from `node::module`](https://nodejs.org/api/module.html#moduleregisterspecifier-parenturl-options), but I'm not entirely sure what all it entails.

**Reply by novemberborn** [reply]
Once it's stable, the next question is whether AVA needs to provide anything for this or if it can be done through the `require` config.

**Comment by JAForbes** [comment]
I managed to get it to work, but it is weird:

```js
// ava.config.js
export default {
	extensions: {
		js: true,
		ts: 'commonjs' // but still write your tests in esm
	}
}
```

```bash
npx tsx node_modules/.bin/ava
```

For some reason `--loader=tsx`, or `--loader=tsx/esm` didn't work, but just letting `tsx` run `ava` solves it.  The `ts:commonjs` feels wrong because my test files are actually using esm, but without setting that, ava didn't automatically find the .ts test files.

I'm guessing tsx is transpiling to cjs which is why `ts: 'commonjs'` has to be set.

I wish this wasn't the solution 😅

**Reply by yuriy-yarosh** [reply]
@JAForbes there are issues in NODE_OPTIONS env var propagation.

It works if passed directly, as 

`NODE_OPTIONS='--import tsx' yarn ava`

The other issue is that tsx doesn't support decorators with esbuild, due to https://github.com/evanw/esbuild/issues/104 
So, I'd consider switching to `ts-loader` but for some reason `--import ts-loader/esm` or similar hangs up ava completely.

**Reply by JAForbes** [reply]
Just gave it a go, `NODE_OPTIONS='import tsx'` works, but the `ts: 'commonjs'` config is still needed.  Definitely an improvement though thank you!

**Reply by Sparticuz** [reply]
After lots of combo's, I've got working tests with the following:
```bash
$ node --version
v20.10.0
```
```json
{
  "type": "module",
  "scripts": {
    "test": "c8 ava"
  },
  "ava": {
    "extensions": {
      "ts": "module"
    },
    "nodeArguments": [
      "--import=tsx/esm"
    ]
  },
"devDependencies" {
    "ava": "5.3.1",
    "c8": "^8.0.1",
    "tsx": "4.2.1",
  }
}
```
ava is pinned because I'm having issues with the non-automatic process closing https://github.com/avajs/ava/issues/3279
tsx is pinned because of the 100% issue https://github.com/privatenumber/tsx/issues/433

**Reply by xfoxfu** [reply]
@Sparticuz I am having difficulties following your setup. Would you please offer your `tsconfig.json` for a reference?

**Reply by Sparticuz** [reply]
Actually, was still having difficulties, but I just switch to tsimp instead of tsx and it seems to be better
```diff
-        "tsx": "^4.2.1",
+        "tsimp": "^2.0.10",
```
Now my only issue is ava not auto-closing

**Reply by codan84** [reply]
Thanks for the above, this worked for me on node 20:
```
{
  cache: false,
  failWithoutAssertions: false,
  extensions: {
    ts: 'module'
  },
  nodeArguments: [
    '--import=tsimp/import',
    '--import=testdouble'
  ]
}
```

**Reply by vashian** [reply]
It works for me, thanks!

**Comment by yuriy-yarosh** [comment]
@Sparticuz fyi TSX is on esbuild, so doesn't support decorators hence TypeORM and MikroORM.

**Comment by RangerMauve** [comment]
I switched my CI to use node 19.x as a quickfix
